### PR TITLE
Analytics Manager, Event Builder, Transforms, Unit Tests

### PR DESCRIPTION
**Pull request by kchitalia**

### CODE REVIEW

**Comment by kchitalia** [comment]
@bhariharan @maliroteh 
Publishing piece coming up soon

File: SalesforceSDK/Analytics/Model/InstrumentationEventBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+            {
+                throw new EventBuilderException("Mandatory field 'event type' not set!");
+            }
+            if (SchemaType != InstrumentationEvent.SchemaType.LightningPerformance && Page == null)
+            {
+                throw new EventBuilderException("Mandatory field 'page' not set!");
+            }
+
+            //TODO: Uncomment the following lines after AnalyticsManager is implemented.
+            var sequenceId = AnalyticsManager.GetGlobalSequenceId() + 1;
```
Comment (bhariharan) [inline SalesforceSDK/Analytics/Model/InstrumentationEventBuilder.cs]: Isn't `AnalyticsManager` implemented?

File: SalesforceSDK/Analytics/Store/EventStoreManager.cs [hunk 2]
```diff
@@ -162,14 +159,6 @@ public EventStoreManager(string fileNameSuffix, string encryptionKey)
             }
         }
 
-        public async Task ChangeEncryptionKeyAsync(string oldKey, string newKey)
```
Comment (bhariharan) [inline SalesforceSDK/Analytics/Store/EventStoreManager.cs]: How are passcode changes handled?

Comment (kchitalia) [inline SalesforceSDK/Analytics/Store/EventStoreManager.cs]: The passcode changes/encryption key changes are handles by the Encryptor class that is implemented in the Universal project.
https://github.com/forcedotcom/SalesforceMobileSDK-Windows/blob/unstable/SalesforceSDK/Universal/Security/Encryptor.cs

**Comment by kchitalia** [comment]
I have added the publishing piece as well. @bhariharan I don't understand why we need this file:

https://github.com/forcedotcom/SalesforceMobileSDK-Android/blob/unstable/libs/SalesforceSDK/src/com/salesforce/androidsdk/analytics/SalesforceAnalyticsManager.java

If you can please explain what's the purpose of this, then I can add in something similar for Windows.

File: SalesforceSDK/Analytics/Store/EventStoreManager.cs (hunk 2 above)
Comment (bhariharan) [inline SalesforceSDK/Analytics/Store/EventStoreManager.cs]: As we discussed offline, the `Encryptor` needs to trigger changing the encryption key used to encrypt the files here. So we'll need to bring back the re-encryption piece here, and add those updates to `Encryptor` to trigger it here.

File: SalesforceSDK/Analytics/Model/InstrumentationEventBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+        private JObject _page;
+        private JObject _previousPage;
+        private JObject _marks;
+
+        private static IAnalyticsManager AnalyticsManager;
+
+        private static IApplicationInformationService ApplicationService =
+            SDKServiceLocator.Get<IApplicationInformationService>();
+
+        //TODO: get connectiontype method
```
Comment (bhariharan) [inline SalesforceSDK/Analytics/Model/InstrumentationEventBuilder.cs]: Add `getConnectionType`?

File: SalesforceSDK/Analytics/Model/InstrumentationEventBuilder.cs [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+            SessionStartTime = SessionStartTime == 0 ? curTime : SessionStartTime;
+            return new InstrumentationEvent(eventId, StartTime, EndTime, Name, Attributes, SessionId,
+                    sequenceId, SenderId, SenderContext, SchemaType, EventType, ErrorType,
+                    deviceAppAttributes, GetConnectionType(), SenderParentId, SessionStartTime, Page,
+                    PreviousPage, Marks);
+        }
+
+        public string GetConnectionType()
+        {
+            return ApplicationService.GetConnectionType();
```
Comment (bhariharan) [inline SalesforceSDK/Analytics/Model/InstrumentationEventBuilder.cs]: Wow - it's that simple? What does this typically return? Does it return subtype too?

File: SalesforceSDK/Analytics/Store/IEventStoreManager.cs [hunk 5]
```diff
@@ -88,15 +88,6 @@ public interface IEventStoreManager
         Task DeleteAllEventsAsync();
 
         /// <summary>
-        /// Changes the encryption key to a new value. Fetches all stored events
-        /// and re-encrypts them with the new encryption key
-        /// </summary>
-        /// <param name="oldKey"></param>
-        /// <param name="newKey"></param>
-        /// <returns></returns>
-        Task ChangeEncryptionKeyAsync(string oldKey, string newKey);
```
Comment (bhariharan) [inline SalesforceSDK/Analytics/Store/IEventStoreManager.cs]: We need to bring this back.

File: SalesforceSDK/Analytics/Model/InstrumentationEventBuilder.cs (hunk 3 above)
Comment (kchitalia) [inline SalesforceSDK/Analytics/Model/InstrumentationEventBuilder.cs]: This is added, I'll remove the TODO

File: SalesforceSDK/Analytics/Model/InstrumentationEventBuilder.cs (hunk 4 above)
Comment (kchitalia) [inline SalesforceSDK/Analytics/Model/InstrumentationEventBuilder.cs]: The actual method is this:
public string GetConnectionType()
        {
            var profile = NetworkInformation.GetInternetConnectionProfile();
            return profile?.ProfileName ?? string.Empty;
        }

File: SalesforceSDK/Universal/Analytics/AILTNPublisher.cs [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+                        var data = new JObject();
+                        var schemaType = eventToPublish[InstrumentationEvent.SCHEMA_TYPE_KEY];
+                        data.Add(InstrumentationEvent.SCHEMA_TYPE_KEY, schemaType);
+                        eventToPublish[InstrumentationEvent.SCHEMA_TYPE_KEY].Remove();
+                        data.Add(PAYLOAD, eventToPublish.ToString());
+                        trackingInfo.Add(DATA, data);
+                        loglines.Add(trackingInfo);
+                    }
+                }
+                body.Add(LOG_LINES, loglines);
```
Comment (bhariharan) [inline SalesforceSDK/Universal/Analytics/AILTNPublisher.cs]: How do we `GZIP` the body of the `POST` here?

**Comment by bhariharan** [comment]
We need to add the `SalesforceAnalyticsManager` class as well, as we discussed offline.

**Comment by kchitalia** [comment]
@bhariharan Added SalesforceAnalyticsManager. In the last PR the method for compressing the request body was missed, I added that as well.

File: SalesforceSDK/Universal/Analytics/SalesforceAnalyticsManager.cs [hunk 7]
```diff
@@ -0,0 +1,313 @@
+﻿using System;
```
Comment (bhariharan) [inline SalesforceSDK/Universal/Analytics/SalesforceAnalyticsManager.cs]: Copyright.

File: SalesforceSDK/Universal/Analytics/SalesforceAnalyticsManager.cs [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+        public async Task PublishEventsAsync(List<InstrumentationEvent> events)
+        {
+            if (events == null || events.Count == 0)
+            {
+                return;
+            }
+            var eventIds = new List<string>();
+            bool success = true;
+            var remoteKeySet = remotes.Keys;
+            foreach (var key in remoteKeySet)
```
Comment (bhariharan) [inline SalesforceSDK/Universal/Analytics/SalesforceAnalyticsManager.cs]: Where is `key` being used?

File: SalesforceSDK/Universal/Analytics/SalesforceAnalyticsManager.cs [hunk 9]
```diff
@@ -0,0 +1,313 @@
...
+            {
+                ITransform transformer = new AiltnTransform();
```
Comment (bhariharan) [inline SalesforceSDK/Universal/Analytics/SalesforceAnalyticsManager.cs]: Shouldn't this iterate through the transforms available to us, a.k.a., `key`?

File: SalesforceSDK/Universal/Analytics/SalesforceAnalyticsManager.cs [hunk 10]
```diff
@@ -0,0 +1,313 @@
...
+                foreach (var instrumentationEvent in events)
+                {
+                    eventIds.Add(instrumentationEvent.EventId);
+                    var eventJson = transformer.Transform(instrumentationEvent);
+                    if (eventJson != null)
+                    {
+                        eventsArray.Add(eventJson);
+                    }
+                }
+                var networkPublisher = new AILTNPublisher();
```
Comment (bhariharan) [inline SalesforceSDK/Universal/Analytics/SalesforceAnalyticsManager.cs]: We need to get all the `value` values for each `key`, a.k.a. each `Publisher` associated with each `Transformer`.

**Comment by kchitalia** [comment]
All tests passing, Iterating over the right items and added the copyright.

**Comment by kchitalia** [comment]
@bhariharan Good to merge this?

File: SalesforceSDK/Universal/Analytics/SalesforceAnalyticsManager.cs [hunk 11]
```diff
@@ -0,0 +1,347 @@
...
+        /// <returns></returns>
+        public async Task PublishEventsAsync(List<InstrumentationEvent> events)
+        {
+            if (events == null || events.Count == 0)
+            {
+                return;
+            }
+            var eventIds = new List<string>();
+            var success = true;
+            var remoteKeySet = remotes.Keys;
```
Comment (bhariharan) [inline SalesforceSDK/Universal/Analytics/SalesforceAnalyticsManager.cs]: @kchitalia Where are you adding the default transformer and publisher, a.k.a. `AILTNTransform` and `AILTNPublisher`, to the list of remotes?

**Comment by bhariharan** [comment]
@kchitalia I added one more comment. Also, can we run an end-to-end test tomorrow to make sure an event is making it into `Splunk` before merging this?

File: SalesforceSDK/Universal/Analytics/SalesforceAnalyticsManager.cs (hunk 11 above)
Comment (kchitalia) [inline SalesforceSDK/Universal/Analytics/SalesforceAnalyticsManager.cs]: @bhariharan They are added in this method:
https://github.com/forcedotcom/SalesforceMobileSDK-Windows/pull/221/files#diff-6159c2f8a1dfec0126976c2ec44670ddR305

**Comment by bhariharan** [comment]
@kchitalia I'm merging this in for now, but let's run an end-to-end `Splunk` test before signing off.

**Pull request merged**